interfaces/devdoc/threadpool_requirements.md

```diff
@@ -104,6 +110,60 @@ MOCKABLE_FUNCTION(, void, threadpool_close, THREADPOOL_HANDLE, threadpool);
 
 **NON_THREADPOOL_01_019: [** If `threadpool` is not OPEN, `threadpool_close` shall return. **]**
 
+### threadpool_create_work_item
+
+```c
+MOCKABLE_FUNCTION(, PVOID, threadpool_create_work_item, THANDLE(THREADPOOL), threadpool, THREADPOOL_WORK_FUNCTION, work_function, PVOID, work_function_context);
+```
+
+`threadpool_create_work_item` creates a work item to be executed by the threadpool.
+
+**NON_THREADPOOL_05_001: [** `threadpool_create_work_item` shall initialize the `work_item_context` to `NULL`, where `NULL` would be a failure indicator to the calling functions **]**
+
+**NON_THREADPOOL_05_002: [** If `threadpool` is `NULL`, `threadpool_create_work_item` shall fail and return a `NULL` value. **]**
+
+**NON_THREADPOOL_05_003: [** If `work_function` is `NULL`, `threadpool_create_work_item` shall fail and return a `NULL` value. **]**
+
+**NON_THREADPOOL_05_004: [** Otherwise `threadpool_create_work_item` shall allocate a context where `work_function`, `work_function_context`, and `ptp_work` shall be saved. **]**
+
+**NON_THREADPOOL_05_005: [** If any error occurs, `threadpool_create_work_item` shall fail and return a `NULL` value. **]**
+
+**NON_THREADPOOL_05_006: [** `threadpool_create_work_item` shall create `PTP_WORK` by calling `CreateThreadpoolWork` to set the callback function as `on_work_callback_v2` and pass the newly created context whenever the callback is called in the tp_environment. **]**
+
+**NON_THREADPOOL_05_007: [** If any error occurs, `threadpool_create_work_item` shall fail, free the newly created context and return a `NULL` value. **]**
+
+### threadpool_schedule_work_item
+
+```c
+MOCKABLE_FUNCTION(, int, threadpool_schedule_work_item, THANDLE(THREADPOOL), threadpool, PVOID, work_item_context);
+```
+
+`threadpool_schedule_work_item` schedules a work item to be executed by the threadpool.
+
+**NON_THREADPOOL_05_008: [** If `threadpool` is NULL, `threadpool_schedule_work_item` shall fail and return a `non-zero` value. **]**
+
+**NON_THREADPOOL_05_009: [** If `work_function` is NULL, `threadpool_schedule_work_item` shall fail and return a `non-zero` value. **]**
+
+**NON_THREADPOOL_05_010: [** If `ptp_work` within the context is `NULL`, `threadpool_schedule_work_item` shall fail and return a `non-zero` value. **]**
```

Comment:
This should not be possible as we should not return a handle in `threadpool_create_work_item` with a `NULL` `ptp_work` and there is no other way to get these handles. So we can't have this happen/can't write a test for it/don't need the spec.

Reply:
These checks are needed in case the variable is accidentally assigned `NULL` after creation of the handles or changed to `NULL` during transmission by a hacker.

Reply:
This is another reason for opaque handles.

No one can change the data fields outside of threadpool_win32.c (or threadpool_linux.c). The type is encapsulated.

Yes, you could cast it to something, modify the memory directly, etc. But we have to assume that the callers to `threadpool_schedule_work_item` are not adversarial. This module is not at a security boundary and isn't going to get random pointer input.

We can't really guard against "hackers" to this call by checking for NULL anyway. What if the field was changed to "1" (or any other invalid value) instead of NULL?